orchagent/fdborch.cpp

```diff
@@ -482,26 +693,69 @@ bool FdbOrch::addFdbEntry(const FdbEntry& entry, const string& port_name, const
     attr.value.s32 = SAI_PACKET_ACTION_FORWARD;
     attrs.push_back(attr);
 
-    if (m_entries.count(entry) != 0) // we already have such entries
+    if(macUpdate)
     {
-        removeFdbEntry(entry);
-    }
+        /* delete and re-add fdb entry instead of update, 
+         * as entry may age out in HW/ASIC_DB before
+         * update, causing the update request to fail.
+         */
+        SWSS_LOG_NOTICE("MAC-Update FDB %s in %s on from-%s:to-%s from-%s:to-%s", entry.mac.to_string().c_str(), vlan.m_alias.c_str(), oldPort.m_alias.c_str(), port_name.c_str(), oldType.c_str(), type.c_str());
+        status = sai_fdb_api->remove_fdb_entry(&fdb_entry);
+        if (status != SAI_STATUS_SUCCESS)
+        { 
+            SWSS_LOG_INFO("FdbOrch RemoveFDBEntry: Failed to remove FDB entry. mac=%s, bv_id=0x%lx",
+                           entry.mac.to_string().c_str(), entry.bv_id);
+        }
```

Comment:
This may call removeFdbEntry(). If sai_fdb_api->create_fdb_entry() return failure, such as m_entries, CRM_FDB_ENTRY counter will mismatched.

Reply:
if removeFdbEntry() fails due to SAI_STATUS_ITEM_NOT_FOUND, then there is no mismatch, as the fdb entry was already aged out.
if removeFdbEntry() fails due to some other reason, then create_fdb_entry() will also fail and return false as its an error condition, its not expected though.

Reply:
It is possible that remove_fdb_entry() success but create_fdb_entry() fail, m_entries, CRM_FDB_ENTRY counter will mismatched in this case.

Reply:
In that case, we log error message return 'false', so it keeps retrying. If it's recoverable, it will pass eventually, otherwise there is nothing much we can do.

Reply:
OK, this is a fairly rare occurrence.

Reply:
Additional example for this case, one MAC is moved from port1 to port2, then it is deleted by application for some reason quickly. As above, remove_fdb_entry() success but create_fdb_entry() fail, this MAC entry is deleted in HW but exist in m_entries. The add event may be overlapped by the following delete event in redis, then only the delete event is need to be handled. In removeFdbEntry(), m_entries.find(entry) may success, and sai_fdb_api->remove_fdb_entry(&fdb_entry) may return failure for the entry is not exist in SAI/HW. If sai_fdb_api->remove_fdb_entry(&fdb_entry) return failure, removeFdbEntry() will return true, and then the delete event is erased by redis. Such codes as 'port.m_fdb_count--' and 'm_entries.erase(entry)' are not executed in removeFdbEntry(), thus m_entries and reference count are inconsistent with SAI/HW forever.

Reply:
if sai_fdb_api->remove_fdb_entry(&fdb_entry) fails, removeFdbEntry() will not return, it will try create_fdb_entry() and if create fails, it returns false. Returning false will cause addFdbEntry to be called again its still in the consumer map. If it keeps failing, there is nothing much we can do, but in theory it shouldn't fail.

Reply:
Sorry, it seems my last words are ambiguous, and may explain more clearly.
1. For example, one MAC is moved from port1 to port2, and then it is deleted by application for some reason quickly. This will cause two events: first is MAC add event, second is MAC delete event.
2. Fdborch first handle MAC add event, call addFdbEntry(). As discussion above, in addFdbEntry(), remove_fdb_entry() success but create_fdb_entry() fail, this MAC entry is deleted in HW but exist in m_entries. And the MAC add event is still in consumer map since addFdbEntry() return false.
3. At this time, MAC delete event comes in redis. Since the key is identical, the MAC delete event will overlap the MAC add event, thus only the MAC delete event is left in consumer map.
4. Fdborch then handle MAC delete event, call removeFdbEntry(). In removeFdbEntry(), m_entries.find(entry) may success, and sai_fdb_api->remove_fdb_entry() may return failure for the entry is not exist in SAI/HW, and removeFdbEntry() will return true. The codes in removeFdbEntry() are:
    status = sai_fdb_api->remove_fdb_entry(&fdb_entry);
    if (status != SAI_STATUS_SUCCESS)
    {
        SWSS_LOG_ERROR("FdbOrch RemoveFDBEntry: Failed to remove FDB entry. mac=%s, bv_id=0x%lx",
                       entry.mac.to_string().c_str(), entry.bv_id);
        return true; //FIXME: it should be based on status. Some could be retried. some not
    }
5. Since removeFdbEntry() return true, the MAC delete event will be erased by redis. In other words, the MAC delete event is removed from consumer map.
6. Such codes as 'port.m_fdb_count--' and 'm_entries.erase(entry)' are behind the code 'return true; //FIXME: it should be based on status. Some could be retried. some not', thus they are not executed in removeFdbEntry(), and thus m_entries and reference count are inconsistent with SAI/HW forever.

Reply:
Got it. Thanks.
I got confused because you mentioned 'event'. mac event comes from SAI, but you are talking about mac add/delete coming from APP_DB.
Yes, we should not return 'true' when mac delete sai call fails, although it is not expected, as the mac added from APP_DB is usually deleted in APP_DB and also a failure will never recover theoretically. I will change this.

Reply:
I still think the source of this issue is in addFdbEntry(), line 710:
status = sai_fdb_api->remove_fdb_entry(&fdb_entry);
        if (status != SAI_STATUS_SUCCESS)
        { 
            SWSS_LOG_INFO("FdbOrch RemoveFDBEntry: Failed to remove FDB entry. mac=%s, bv_id=0x%lx",
                           entry.mac.to_string().c_str(), entry.bv_id);
        }
When we call sai_fdb_api->remove_fdb_entry() success to delete the MAC from SAI/HW, we must remove the item from m_entries and decrease the reference count at the same time(Before call sai_fdb_api->create_fdb_entry(), since sai_fdb_api->create_fdb_entry() may return failed).
I think change 'return true' to 'return false' in removeFdbEntry() is not helpful, since the MAC is deleted from SAI/HW, and sai_fdb_api->remove_fdb_entry() will return failed forever, and this may cause dead loop.

Reply:
If we clean up the fdb in m_entries and then sai_fdb_api->create_fdb_entry() fails, it will immediately add it back when createFdbEntry() is retried. So, either we leave it this way or we should not return false even if sai_fdb_api->create_fdb_entry() returns failure. But, there are cases like failure due to hash collision which may eventually pass if retried. What so you think?

Reply:
Yes, If we clean up the fdb in m_entries and then sai_fdb_api->create_fdb_entry() fails, it may immediately or after some time add it back when createFdbEntry() is retried(It also may be never success). Logically, first remove then add to m_entries, it is not wrong. If we not do this, in the case I list 5 days ago(6 steps, application such as ICCP move one MAC from port1 to port2, and then delete it for some reason quickly), m_entries and reference count will inconsistent with SAI/HW forever.

Reply:
sounds good. i will make this change.